libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,616 @@
+//	  (dgram->destination_nodes_len - store->rx_bytes_read) : CAN_BUFFER_SIZE;
+//  if (!prv_read_rx_buffer((dgram->destination_nodes + store->rx_bytes_read), rx_len)) {
+//    event_raise_no_data(store->repeat_event);
+//  } else {
+//    store->rx_bytes_read += rx_len;
+//    LOG_DEBUG("RX LEN: %zu, %d\n", store->rx_bytes_read, dgram->destination_nodes_len);
+//    if (store->rx_bytes_read == dgram->destination_nodes_len) {
+//      LOG_DEBUG("data_len\n");
+//      event_raise_no_data(store->rx_event);
+//      store->rx_bytes_read = 0;
+//    } else {
+//      event_raise_no_data(store->repeat_event);
+//    }
+//  }
+  if (!prv_read_rx_buffer(dgram->destination_nodes, dgram->destination_nodes_len)) {
+    event_raise_no_data(store->repeat_event);
+  } else {
+      LOG_DEBUG("data_len\n");
+      event_raise_no_data(store->rx_event);
+  }   
+} 
+
+
+static void prv_process_data_len_rx(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  CanDatagram *dgram = &store->dgram;
+  if (!prv_read_rx_buffer((uint8_t*)&dgram->data_len, DATA_LEN_SIZE_BYTES)) {
+    event_raise_no_data(store->repeat_event);
+  } else {
+    event_raise_no_data(store->rx_event);
+    LOG_DEBUG("data\n");
+  }
+}
+
+static void prv_process_data_rx(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  CanDatagram *dgram = &store->dgram;
+//  size_t rx_len = (dgram->data_len - store->rx_bytes_read) < CAN_BUFFER_SIZE ?
+//       (dgram->data_len - store->rx_bytes_read) : CAN_BUFFER_SIZE;
+//  // If read fails, it means all data is not yet available in fifo - call again
+//  if (!prv_read_rx_buffer((dgram->data + store->rx_bytes_read), rx_len)) {
+//    event_raise_no_data(store->repeat_event);
+//  } else {
+//    store->rx_bytes_read += rx_len;
+//    if (store->rx_bytes_read == dgram->data_len) {
+//      if (prv_can_datagram_compute_crc() == dgram->crc) {
+//        event_raise_no_data(store->rx_event);
+//        LOG_DEBUG("datagram complete!\n");
+//      } else {
+//        LOG_WARN("CRC VALUE DID NOT MATCH. EXITING WITH ERROR...\n");
+//        event_raise_no_data(store->error_event);
+//        store->rx_bytes_read = 0;
+//      }
+//    } else {
+//      event_raise_no_data(store->repeat_event);
+//    }
+//  }
+  if (!prv_read_rx_buffer(dgram->data, dgram->data_len)) {
+    event_raise_no_data(store->repeat_event);
+  } else {
+    if (prv_can_datagram_compute_crc() == dgram->crc) {
+      event_raise_no_data(store->rx_event);
+      LOG_DEBUG("datagram complete!\n");
+    } else {
+      LOG_WARN("CRC VALUE DID NOT MATCH. EXITING WITH ERROR...\n");
+      event_raise_no_data(store->error_event);
+      LOG_DEBUG("data_len\n");
+      event_raise_no_data(store->rx_event);
+    }
+  }
+}
+ 
+static void prv_process_done(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  // Handle tx specific cleanup
+  if (e->id == store->tx_event) {
+    //tx any final data
```

Comment:
This is valid, as the max length pushed to the fifo for tx is 8 bytes, so we won't lose any data